### ISSUE TITLE
Is there any way to build as static engine for OpenSSL

### ISSUE THREAD

**Issue by kartaris** [gost-engine/engine#340]
I want to include OpenSSL with gost-engine into an iOS application as a framework and there is no way to include dynamic libraries there.
OpenSSL 1.1.1

**Comment by beldmit** [comment]
There is no direct way to do it. LibreSSL includes some sort of GOST implementation, so it may be an option.

**Comment by kartaris** [comment]
> There is no direct way to do it. LibreSSL includes some sort of GOST implementation, so it may be an option.

Thanks for your reply. 
I tried libressl and I get [tlsv1 alert decode error](https://github.com/libressl-portable/portable/issues/679#issuecomment-895115869) when making request via cURL. Maybe you have an idea how to fix it?

**Comment by kartaris** [comment]
I added some more information to the issue libressl-portable/portable#679.
And I've found similar issue #310 here. May they be connected somehow?

**Comment by beldmit** [comment]
I'm not familiar with LibreSSL beyond the information I've provided. For me it looks like semiabandoned fork of OpenSSL with some sort of features added. As they don't like engines, they accepted the GOST algorithms implementation into the main tree.

#310 looks like an endianess problem to me - if the correct endianess is not set, we get incorrect hash value. The native GOST engine build detects the endianess.

@lumag if you remember anything related to LibreSSL, please let us know :)

**Comment by kartaris** [comment]
> I'm not familiar with LibreSSL beyond the information I've provided. For me it looks like semiabandoned fork of OpenSSL with some sort of features added. As they don't like engines, they accepted the GOST algorithms implementation into the main tree.
> 
> #310 looks like an endianess problem to me - if the correct endianess is not set, we get incorrect hash value. The native GOST engine build detects the endianess.
> 
> @lumag if you remember anything related to LibreSSL, please let us know :)

Is there any documentation about including the engine logic to the main tree?
Think it would a good solution for me. 
I can start working on it while waiting for the answer from LibreSSL developers(I'm afraid waiting can take much more time than doing something on my own)

**Comment by lumag** [comment]
If I remember correctly LibreSSL has not been updated to support recent crypto standards. In the past I've provided one of local vendors with the patchset to update it, but the upstreaming process stopped because of the lack of reviews from the LibreSSL side.

I don't remember any particular issues from the process of merging the GOST engine into the source tree. It has been pretty straightforward.

**Comment by kartaris** [comment]
> If I remember correctly LibreSSL has not been updated to support recent crypto standards. In the past I've provided one of local vendors with the patchset to update it, but the upstreaming process stopped because of the lack of reviews from the LibreSSL side.
> 
> I don't remember any particular issues from the process of merging the GOST engine into the source tree. It has been pretty straightforward.

I'm not in touch with openssl library, it's the first experience in my practice in integrating something into openssl. Do I need to make some bindings to integrate the GOST logic there?

**Comment by kartaris** [comment]
> > I don't remember any particular issues from the process of merging the GOST engine into the source tree. It has been pretty straightforward.
> 
> I'm not in touch with openssl library, it's the first experience in my practice in integrating something into openssl. Do I need to make some bindings to integrate the GOST logic there?

Or it would be enough if I change gost-engine's CMakeLists.txt to build static lib instead of shared and call ENGINE_load_gost to configure openssl in runtime?

**Comment by beldmit** [comment]
Sorry, can't help you much
For your particular purposes I'd copied the gost files to `crypto/gost` folder of openssl, and added a makefile similar to one in, say, `crypto/rsa`. It's a lot of work, BTW

**Comment by kartaris** [comment]
> Or it would be enough if I change gost-engine's CMakeLists.txt to build static lib instead of shared and call ENGINE_load_gost to configure engine in runtime?

I just finished building engine as static library and it works now. But there is an issue when building with gcc: libgost_core depends on gost and gost depends on libgost_core. That's why I had to link gost_core twice like on the 

![image](https://user-images.githubusercontent.com/10552394/129024178-5b9d9003-c3fb-403d-afeb-7ff38d25b947.png)

I think libgost_core must be fully independend.

PS: dynamic link doesn't make sense to it, only static does.

**Comment by beldmit** [comment]
If you make it independent, I'll accept the patch.

**Comment by kartaris** [comment]
> If you make it independent, I'll accept the patch.

Ok. I will try. Should I add static build to the patch?

**Comment by levitte** [comment]
How about trying this simple patch?

``` diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 958f48f..861a230 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -340,7 +340,7 @@ target_link_libraries(gost_engine PRIVATE gost_core gost_err
                       ${OPENSSL_CRYPTO_LIBRARY})
 
 # The GOST engine in library form
-add_library(lib_gost_engine SHARED ${GOST_ENGINE_SOURCE_FILES})
+add_library(lib_gost_engine STATIC ${GOST_ENGINE_SOURCE_FILES})
 set_target_properties(lib_gost_engine PROPERTIES
   COMPILE_DEFINITIONS "BUILDING_ENGINE_AS_LIBRARY"
   PUBLIC_HEADER gost-engine.h
```

**Comment by kartaris** [comment]
> How about trying this simple patch?
> 
> ```diff
> diff --git a/CMakeLists.txt b/CMakeLists.txt
> index 958f48f..861a230 100644
> --- a/CMakeLists.txt
> +++ b/CMakeLists.txt
> @@ -340,7 +340,7 @@ target_link_libraries(gost_engine PRIVATE gost_core gost_err
>                        ${OPENSSL_CRYPTO_LIBRARY})
>  
>  # The GOST engine in library form
> -add_library(lib_gost_engine SHARED ${GOST_ENGINE_SOURCE_FILES})
> +add_library(lib_gost_engine STATIC ${GOST_ENGINE_SOURCE_FILES})
>  set_target_properties(lib_gost_engine PROPERTIES
>    COMPILE_DEFINITIONS "BUILDING_ENGINE_AS_LIBRARY"
>    PUBLIC_HEADER gost-engine.h
> ```

I've done the same in my PR. I just added a way to build engine statically without breaking the ability to load it as dynamic engine for OpenSSL. That's why I kept SHARED library building.

**Comment by kartaris** [comment]
Fixed in 1.1
